### PR TITLE
130 time picker

### PR DESCRIPTION
**Pull request by IngaSokman**

### CODE REVIEW

**Comment by KaarelKa** [comment]
Seems to have some merge conflicts now, should be synced with dev branch

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [inline src/components/atoms/TimeColumn/TimeColumn.tsx]: **[blocker]** this shouldn't be an atom. Basically the logic is that atoms can't consist of other atoms. If we have a component that consists of atoms, then that should already be a molecule and this one uses 2 atoms already

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [inline src/components/atoms/TimeColumn/TimeColumn.tsx]: **[blocker]**

Let's stick with absolute imports for cases like this.

Relative imports for components should be used, when importing some smaller part of this component and only if that part is only used inside this component + is situated in the same component folder (basically if that part will always be moved with this component).

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [inline src/components/atoms/TimeColumn/TimeColumn.tsx]: **[blocker]**

If I'm not mistaken, then code will never reach this default return.  "selectedTime !== end - 1" covers all cases, other than "selectedTime === end - 1" so there will be no condition left, which would pas by both "if" statements

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [inline src/components/atoms/TimeColumn/TimeColumn.tsx]: **[blocker]**

Every text we use should be translatable. If we keep these hardcoded strings inside the codebase, it will be a pain to find them later on. Only time these are allowed, is when the component/feature is WIP.

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [inline src/components/atoms/TimeColumn/TimeColumn.tsx]: **[blocker]**

No need for else after return, however here returning sth doesn't really make sense, so rather keep the else and remove the returns

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [reply src/components/atoms/TimeColumn/TimeColumn.tsx]: Also returning the call of a function here is not really necessary. We just need to call out the function, not return whatever the call of that function returns

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [inline src/components/atoms/TimeColumn/TimeColumn.tsx]: **[blocker]** We have added classNames library for adding conditional classes. Let's use it, instead of constructing the string manually, to keep the codebase consistent

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [inline src/components/atoms/TimeColumn/TimeColumn.tsx]: **[blocker]**

Possibly - Disccuss this with Martine to make sure, but to me it seems that up arrow should increase the number and down arrow should decrease the number, not the other way around.

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [inline src/components/atoms/TimeColumn/TimeColumn.tsx]: **[blocker]** Do we need 2 containing divs here with just 1 child ? In most cases this is not needed, unless you are doing sth very specific. In this case I think we can easily skip 1 of them

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/DatePickerInput.tsx]: **[blocker]**

Let's split Datepicker and Timepicker into 2 separate inputs with their own types. They have very little in common and this will be quite hard to maintain. 

However you can create a new helper component out of the shared part (content container, label, wrapper and errormessage). Both Timepicker and Datepicker can then use that helper component as a wrapper and we will avoid code duplication without ending up with a messy component

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/DatePickerInput.tsx]: **[blocker]**

Currnently clicking on the label of any time or date input on the page will result in the first date picker of the page being focused. Every label should be attached the the corresponding input and should only focus that input.

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/DatePickerInput.tsx:4]: **[blocker]** new console error added with timepicker, always check for new console errors/warning before creating a PR and fix them.


Screenshot 2023-05-15 at 10 43 19

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [reply src/components/molecules/DatePickerInput/DatePickerInput.tsx]: Will review these again after the split, currently not checking any further

File: src/components/molecules/DatePickerInput/styles.module.scss
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/styles.module.scss]: **[blocker]** Generally padding and margin values should use "spacing", however in this case you are using padding to center the label. We should instead use layout css for this. Currently this is a flex container with direction row, so align-items should be used

Comment (KaarelKa) [reply src/components/molecules/DatePickerInput/styles.module.scss]: We do this, because it will be much easier to make the components responsive once the time comes. It will mean less changes, if we for example have different font-size/line-height/input height on smaller screens or they are just changed some time in the future

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/DatePickerInput.tsx]: **[blocker]** seems that accessibility support is not great for these either. Every component should support interaction with just the keyboard as well. Currently timepicker can't be opened with just the keyboard

File: src/components/molecules/TimePickerInput/TimePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/TimePickerInput/TimePickerInput.tsx]: **[blocker]** forwardRef not used correctly here. You are wrapping TimeInput with forwardRef, however the forwarded ref is not used. Instead the ref is passed as a prop (inputRef) and passed from there. Let's stick with using forwardRef, not ref as a prop.

File: src/components/molecules/TimePickerInput/TimePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/TimePickerInput/TimePickerInput.tsx]: **[blocker]** I think we are using too many internal states currently. hour, minute and second value are currently kept in 3 places at once (Inside TimePickerInput, inside the form and inside TimeColumn).

Ideally we want to keep the value only in 1 place to avoid issues that come from data duplication. In this case it is kind of okay to have it in 2 places, but 3 seems overkill

File: src/components/molecules/TimePickerInput/TimePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/TimePickerInput/TimePickerInput.tsx]: **[blocker]**

We will most likely use outside click logic in many places in this project, so would be good to abstract this away or use some 3rd party hook for it, so we can reuse the logic easily. 1 option would be to use https://ahooks.js.org/hooks/use-click-away 

Extra note is that currently you are only handling mouse clicks, which means that this won't work for people who are using a keyboard instead of a mouse. Support for this can be added quite easily with https://ahooks.js.org/hooks/use-click-away or here, by just adding a few extra event listeners

File: src/components/molecules/TimePickerInput/TimePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/TimePickerInput/TimePickerInput.tsx]: **[blocker]**

Naming wise it would be good to use verbs for function names. It will make the codebase more predictable and easier to read (every verb does something, which means they are some type of method/function)

File: src/components/molecules/TimePickerInput/TimePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/TimePickerInput/TimePickerInput.tsx]: **[blocker]** Can't pick 59 minutes or 59 seconds right now

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/DatePickerInput.tsx]: **[blocker]** a lot of duplicated types here. Everything that is defined under DatePickerComponentProps is also defined under DatePickerInputProps, so DatePickerInputProps could just extend DatePickerComponentProps.

Only exception is ariaLabelToUse, but this is not exactly necessary and is also not correctly defined. It is not required by DatePickerComponent, however it is always passed, since ariaLabel is required by DatePickerInput. There is also no need to pass both ariaLabel and ariaLabelToUse here, you will not need both, so just ariaLabel will be enough.

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/DatePickerInput.tsx]: **[blocker]** icons don't need aria labels. aria-label is meant for interactive elements: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label

File: src/components/molecules/TimePickerInput/TimePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/TimePickerInput/TimePickerInput.tsx]: **[blocker]**

Let's remove some type repetition from here as well. It can be quite confusing, since it's hard to tell what props have the same type between parent and child (which ones are passed directly from parent to child) and which ones are new/modified

**Review by KaarelKa — CHANGES_REQUESTED** [review]

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (IngaSokman) [reply src/components/atoms/TimeColumn/TimeColumn.tsx]: Yes, true, this default return is redundant, removed it

File: src/components/atoms/TimeColumn/TimeColumn.tsx
Comment (IngaSokman) [reply src/components/atoms/TimeColumn/TimeColumn.tsx]: Good point, asked Martine just to be sure, waiting for reply currently

Comment (IngaSokman) [reply src/components/atoms/TimeColumn/TimeColumn.tsx]: It should be the other way around - top arrow buttons increase the number and down arrows decrease the number, changed it

File: src/components/atoms/Icon/Icon.tsx
Comment (KaarelKa) [inline src/components/atoms/Icon/Icon.tsx]: **[blocker]**

Hmm I'm wondering what is the point of this wrapper component. This seems to be meant for icons as svgs, however those already have the className prop. Se essentially this component doesn't really seem to do anything.
It currently just add 1 extra import + some extra complexity, but no real value.

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/DatePickerInput.tsx]: **[non-blocker]**

Seems a bit unnecessary to create a new variable here, it is essentially const a = b

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/DatePickerInput.tsx:22]: **[blocker]**

These seem to be duplicated types. In reality DatePickerInputProps seem to be a combination of SharedProps (or rather DatePickerComponentProps) and InputWrapperProps. You could import InputWrapperProps here instead of duplicating. This would also make it a lot clearer for other developers that the parent component (DatePickerInput) accepts props for InputWrapper and DatePickerComponent.

Only prop inside InputWrapperProps, which is not here is children, however that can just be omitted

File: src/components/molecules/DatePickerInput/DatePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/DatePickerInput/DatePickerInput.tsx:94]: **[blocker]**

Previously we had the issue where clicking on the label always opened the first DatePicker, but now clicking on the label doesn't open any datepicker.

Basically for all input fields, clicking on the label should focus on the input. So for datepicker it should open the datepicker.

This is built into html as well. If you have any input component with some id, then you can bind any label component to it. https://www.w3schools.com/tags/att_label_for.asp

Basically you can bind them like this:

<label htmlFor="randomInput">text</label>
<input id="randomInput" />

For other inputs (Like checkbox or TextInput, this works because, because @radix-ui/react-form does this automatically. their <Label> and <Control> elements are bound, however we are not using Control here, so this should be done manually

File: src/components/molecules/TimeColumn/TimeColumn.tsx
Comment (KaarelKa) [inline src/components/molecules/TimeColumn/TimeColumn.tsx]: **[blocker]**

Would prefer a different implementation here.

1. Effectively we are just changing the number, so there is no need to create so many divs. Currently it is creating 60 separate divs for minutes, when we could just have 1 div tag, which displays value formatted correctly.

2. All text should be in appropriate tags. divs should never contain text directly, we have a list of htmls tags meant specifically for different types of text in different contexts

3. Can ask from Martine about this, but when interacting with this input I find myself wanting to use by keyboard to write the numbers in, instead of pressing the arrow button 30 times (if I want to select 16:30 time for example), so I would make these into input fields. We should still have the arrows, but the option to edit the number manually would make it much more user friendly.

File: src/components/molecules/TimePickerInput/TimePickerInput.tsx
Comment (KaarelKa) [inline src/components/molecules/TimePickerInput/TimePickerInput.tsx:156]: **[blocker]**

Same issue as with DatePicker, can't focus, when pressing on label

**Review by KaarelKa — CHANGES_REQUESTED** [review]

**Review by KaarelKa — APPROVED** [review]

**Comment by KaarelKa** [comment]
@IngaSokman Approved, but has merge conflict. You can merge this after fixing the conflicts

**Pull request merged**
